### PR TITLE
[Nereids](Variant) Implement variant type and support new sub column access method

### PR DESCRIPTION
**Pull request by eldenmoon**
## Proposed changes

The query SELECT v["a"]["b"] from simple_var WHERE cast(v["a"]["b"] as int) = 1

1. During the binding stage, the expression element_at(var, "xxx") is transformed into a SlotReference with a specified path. This conversion is tracked in the StatementContext, where the parent slot is the primary key and the paths are secondary keys. This structure, known as subColumnSlotRefMap in the StatementContext, helps to eliminate duplicates of the same slot derived from identical paths.
2. A new rule, BindSlotWithPaths, is introduced in the analysis stage. This rule is responsible for converting slots with paths into their respective slot suppliers. To ensure that slots with paths are correctly associated with the appropriate LogicalOlapScan, an additional mapping, slotToRelation, is added to the StatementContext. This mapping links the top-level slot to its corresponding relation (i.e., LogicalOlapScan). Consequently, subsequent slots with paths can determine the correct LogicalOlapScan to merge with and modify accordingly.

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]
LGTM

File: regression-test/suites/variant_p0/column_name.groovy [hunk 1]
```diff
@@ -28,19 +28,21 @@ suite("regression_test_variant_column_name", "variant_type"){
         properties("replication_num" = "1", "disable_auto_compaction" = "true");
     """ 
 
+    // sql "set experimental_enable_nereids_planner = false"
+
     sql """insert into ${table_name} values (1, '{"中文" : "中文", "\\\u4E2C\\\u6587": "unicode"}')"""
     qt_sql """select v:中文, v:`\\\u4E2C\\\u6587` from ${table_name}"""
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply regression-test/suites/variant_p0/column_name.groovy:34]: next PR will change

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/Project.java [hunk 2]
```diff
@@ -64,6 +68,22 @@ default List<NamedExpression> mergeProjections(Project childProject) {
         return PlanUtils.mergeProjections(childProject.getProjects(), getProjects());
     }
 
+    /**
+     * Check if it is a project that is pull up from scan in analyze rule
+     * e.g. BindSlotWithPaths
+     */
+    default boolean isPulledUpProjectFromScan() {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/Project.java:75]: hasPulledUpProjectFromScan is more accurate

**Review by xiaokang — APPROVED** [review]
LGTM

**Comment by eldenmoon** [comment]
comments according to CR will be modified in next PR soon

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 3]
```diff
@@ -49,8 +49,11 @@ public enum RuleType {
     BINDING_UNBOUND_TVF_RELATION_FUNCTION(RuleTypeClass.REWRITE),
     BINDING_SET_OPERATION_SLOT(RuleTypeClass.REWRITE),
     BINDING_INLINE_TABLE_SLOT(RuleTypeClass.REWRITE),
-
     COUNT_LITERAL_REWRITE(RuleTypeClass.REWRITE),
+    BINDING_SLOT_WITH_PATHS_PROJECT(RuleTypeClass.REWRITE),
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:53]: done

**Pull request merged**